### PR TITLE
Builtin Zenoh Source and Sink

### PR DESCRIPTION
**Pull request by gabrik**
## Rationale

In order to have tighter integration with Zenoh, built-in sources and sinks that can subscribe/publish directly on Zenoh are needed.

*Note: In the future more builtin Source/Sink could be added (i.e., DDS, ROS, ROS 2, ...)* 

## Code organization

The `builtin` package will be added to `runtime/dataflow/instance` it will contain the implementation of the `Source` and `Sink` traits for `ZenohSink` and `ZenohSource`.

Both source and sink configurations will allow configuring the `key expressions` to subscribe/publish and map between the inputs/outputs and the `ke.`

## Impact on the data flow descriptor

This feature will have minimal impact on the data flow descriptor.
Here an example of a data flow descriptor that uses the builtin Sources and Sink.

```
flow: hello-world

sources:
  - id: zenoh-sub
    configuration:
	key-expressions: 
	    id1: some/key/expr/to/subscribe
	    id2: some/key/expr/to/subscribe2
	    id3: some/key/expr/to/subscribe3
    descriptor: builtin://zenoh/source


operators:
  - id: greetings-maker
    descriptor: file://./nodes/rust/greetings-maker/greetings-maker.yaml


sinks:
  - id: zenoh-pub
    configuration:
        key-expressions: 
	    id1: some/key/expr/to/publish
	    id2: some/key/expr/to/publish2
    descriptor: builtin://zenoh/sink

links:
  - from:
      node: zenoh-sub
      output: id1
    to:
      node: greetings-maker
      input: name

  - from:
      node: greetings-maker
      output: greeting
    to:
      node: zenoh-pub
      input: id1
```

### CODE REVIEW

File: zenoh-flow/src/model/mod.rs [hunk 1]
```diff
@@ -12,6 +12,43 @@
...
+        "zenoh".to_string()
+    }
+}
+
+#[derive(Debug)]
+/// Zenoh-Flow's custom URI struct used for loading nodes.
+pub(crate) enum URIStruct {
+    File(PathBuf),
+    Builtin(Middleware, NodeKind),
+}
```
Comment (J-Loudet) [inline zenoh-flow/src/model/mod.rs:54]: I'd suggest the following:

```suggestion
/// The URI supported by Zenoh-Flow.
pub(crate) enum Uri {  // or possibly ZFUri
    File(PathBuf),
    Builtin(Middleware),
}
```

1. Calling an enum `Struct` feels weird; if we need to separate between Zenoh-Flow's URI and the uri from the package we can qualify the path.
2. The `NodeKind` is redundant with the declaration and makes our code heavier (error cases to handle).

File: zenoh-flow/src/model/mod.rs [hunk 2]
```diff
@@ -12,6 +12,43 @@
...
+
+impl FromStr for Middleware {
+    type Err = ZFError;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let s = s.to_lowercase();
+        if s == "zenoh" {
+            return Ok(Self::Zenoh);
+        }
+        bail!(ErrorKind::ParsingError, "{s} is not a valid middleware!")
```
Comment (J-Loudet) [inline zenoh-flow/src/model/mod.rs:39]:
```suggestion
        bail!(ErrorKind::ParsingError, "Unsupported middleware: '{s}'. Currently supported middlewares: 'zenoh'.")
```

Or a similar message that lists the supported middleware.

File: zenoh-flow/src/model/descriptor/node/mod.rs [hunk 3]
```diff
@@ -154,20 +168,45 @@ impl NodeDescriptor {
...
+                                ErrorKind::MissingConfiguration,
+                                "Builtin Zenoh Sink needs a configuration!"
+                            )
+                        }
+                    },
+                    NodeKind::Sink => {
+                        bail!(ErrorKind::Unimplemented, "Cannot load an sink as source")
+                    }
+                },
+            },
```
Comment (J-Loudet) [inline zenoh-flow/src/model/descriptor/node/mod.rs:209]: It feels like `kind` is redundant and introduces possibilities for errors (as shown by the 3 branches of the`match`).
If a builtin node is declared under the `sources` section then it's a source.

Did I miss anything?

File: zenoh-flow/src/model/descriptor/node/mod.rs [hunk 4]
```diff
@@ -181,37 +220,58 @@ impl NodeDescriptor {
...
+                                ErrorKind::MissingConfiguration,
+                                "Builtin Zenoh Sink needs a configuration!"
+                            )
+                        }
+                    },
+                    NodeKind::Source => {
+                        bail!(ErrorKind::Unimplemented, "Cannot load an source as sink")
+                    }
+                },
+            },
```
Comment (J-Loudet) [inline zenoh-flow/src/model/descriptor/node/mod.rs:260]: Same comment as for the built-in Source, `kind` feels redundant.

File: zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+        for s in &self.subscribers {
+            let (id, sub) = s;
+            futs.push(Box::pin(wait_zenoh_input(id.clone(), sub)))
+        }
+
+        // TODO:
+        // Let's get the first one that finished and send the result to the
+        // corresponding output.
+        // The ones that did not finish are stored and used in next
+        // iteration.
```
Comment (J-Loudet) [inline zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs:173]: What's blocking for this to be implemented?

File: zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                        .ok_or(zferror!(
+                            ErrorKind::ConfigurationError,
+                            "Unable to convert value to string: {:?}",
+                            value
+                        ))?
+                        .to_string();
+
+                    let input = inputs.take_raw(id).ok_or(zferror!(
+                        ErrorKind::MissingOutput(id.clone()),
+                        "Unable to find output: {id}"
```
Comment (J-Loudet) [inline zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs:283]:
```suggestion
                        ErrorKind::MissingInput(id.clone()),
                        "Unable to find input: {id}"
```

File: zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+        for i in &self.inputs {
+            let (id, input) = i;
+            futs.push(Box::pin(wait_flow_input(id.clone(), input)))
+        }
+
+        // TODO:
+        // Let's get the first one that finished and send the result to the
+        // corresponding output.
+        // The ones that did not finish are stored and used in next
+        // iteration.
```
Comment (J-Loudet) [inline zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs:325]: What's blocking?

**Review by J-Loudet — CHANGES_REQUESTED** [review]

File: zenoh-flow/src/utils.rs [hunk 8]
```diff
@@ -12,34 +12,60 @@
...
+use crate::model::{Middleware, ZFUri};
 use crate::prelude::ErrorKind;
 use crate::{bail, zferror, Result};
 use std::path::{Path, PathBuf};
+use std::str::FromStr;
 use url::Url;
 
 /// Given a string representing a [`Url`](`url::Url`), transform it into a
-/// [`PathBuf`](`std::path::PathBuf`).
+/// [`URIStruct`](`zenoh_flow::runtime:dataflow::loader::URIStruct`).
```
Comment (J-Loudet) [inline zenoh-flow/src/utils.rs:23]:
```suggestion
/// [`ZFUri`](`zenoh_flow::model::ZFUri`).
```

(not sure about the path)

File: zenoh-flow/src/utils.rs [hunk 9]
```diff
@@ -12,34 +12,60 @@
...
 ///
 /// # Errors
 ///
 /// This function will return an error in the following situations:
 /// - The provided string does not match the syntax of a [`Url`](`url::Url`).
-/// - The scheme used for the url is NOT `file://`.
-/// - The resulting path cannot be [`canonicalized`](`std::fs::canonicalize`).
-pub(crate) fn try_make_file_path(url_str: &str) -> Result<PathBuf> {
+/// - The scheme is not supported
+/// - In case of `builtin://`, the URI struct does not match `<middleware>/[source|sink]`
```
Comment (J-Loudet) [inline zenoh-flow/src/utils.rs:34]:
```suggestion
/// - In case of `builtin://`, the authority part, `<middleware>`, is not supported
```

**Review by J-Loudet — CHANGES_REQUESTED** [review]

File: zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+                // Add all the inputs during first iteration
+                None => state
+                    .remaining
+                    .push(Box::pin(wait_flow_input(id.clone(), input.clone()))),
+                // Add only the last one that finished as new future.
+                Some(last_id) => {
+                    if id == last_id {
+                        state
+                            .remaining
+                            .push(Box::pin(wait_flow_input(id.clone(), input.clone())));
```
[the remark this answers is not in the document]
Comment (gabrik) [reply zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs:368]: Nope life time issues:
```
error: lifetime may not live long enough
   --> zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs:364:35
    |
345 |     async fn iteration(&self) -> ZFResult<()> {
    |                        - lifetime `'life0` defined here
...
364 |                             .push(Box::pin(async {(id.clone(), input.clone().recv().await) }));
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cast requires that `'life0` must outlive `'static`

```

File: zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+                    id.clone(),
+                    input.receiver.clone(),
+                ))),
+                // Add only the last one that finished as new future.
+                Some(last_id) => {
+                    if id == last_id {
+                        state.remaining.push(Box::pin(wait_zenoh_input(
+                            id.clone(),
+                            input.receiver.clone(),
+                        )));
```
[the remark this answers is not in the document]
Comment (gabrik) [reply zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs:188]: Same lifetime issue

File: zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+        let mut state = self.state.lock().await;
+
+        // Iterate over the intputs
+        for i in &self.inputs {
+            let (id, input) = i;
+            match &state.last {
+                // Add all the inputs during first iteration
+                None => state
+                    .remaining
+                    .push(Box::pin(wait_flow_input(id.clone(), input.clone()))),
```
[the remark this answers is not in the document]
Comment (gabrik) [reply zenoh-flow/src/runtime/dataflow/instance/builtin/zenoh.rs:362]: Same as below

**Pull request merged**